app/vep/models/upload_vcf_files.py

```diff
@@ -0,0 +1,82 @@
+"""
+See the NOTICE file distributed with this work for additional information
+regarding copyright ownership.
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+"""
+import os
+import logging
+import tempfile
+import shutil
+
+from starlette.requests import ClientDisconnect
+
+from streaming_form_data import StreamingFormDataParser
+from streaming_form_data.targets import FileTarget, ValueTarget
+
+from core.config import VCF_UPLOAD_DIRECTORY
+
+MAX_FILE_SIZE = 260 * 1024 * 1024  # 250 MB
+MAX_REQUEST_BODY_SIZE = MAX_FILE_SIZE + 1024
+
+class MaxBodySizeException(Exception):
+    def __init__(self, body_len: str):
+        self.body_len = body_len
+
+
+class MaxBodySizeValidator:
+    def __init__(self, max_size: int):
+        self.body_len = 0
+        self.max_size = max_size
+
+    def __call__(self, chunk: bytes):
+        self.body_len += len(chunk)
+        if self.body_len > self.max_size:
+            raise MaxBodySizeException(body_len=self.body_len)
+
+
+class Streamer:
+    def __init__(self, request):
+        self.request = request
+        self.filename = self.request.headers.get('Filename', None)
```

Comment:
Is this an entirely custom header name, or did you see it used somewhere?

It seems that the standard header for this sort of thing is Content-Disposition. See the [article](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#as_a_header_for_a_multipart_body) on MDN:

![image](https://github.com/Ensembl/ensembl-web-tools-api/assets/6834224/a44dc646-9dd2-4c7e-bdde-40db98e58aa4)

The way that the article is phrased convinced me that you were right asking for the headers.

Reply:
We looked through the docs and the request payloads and It turns out `Content-Disposition` headers with the original filename are already included by the browser in each POST payload, but these headers are part of the form body itself, so would need the addition of the renaming action in serer side like this:
`os.rename(self.filepath, os.path.join(self.temp_dir, self._input_file.multipart_filename))`
But the issue with it is that it takes (some amount of) time? Sorry to come back to this but do we know if/how much time the renaming would actually take? Surely it must be faster than streaming/writing bytes to the file?

Reply:
@veidenberg 

> Sorry to come back to this but do we know if/how much time the renaming would actually take? 

it takes around 380ms to rename 221MB file

```
$ ls -lh original_file 
-rw-r--r-- 1 6630 users 231M Mar 22  2023 original_file

$ cat rename_duration.py 
import os
src="/usr/data/original_file"
dst="/usr/data/renamed_file"
os.rename(src,dst)

$ time python rename_duration.py 
real	0m0.038s
user	0m0.012s
sys	0m0.006s

$ ls -lh renamed_file 
-rw-r--r-- 1 6630 users 231M Mar 22  2023 renamed_file
```

Reply:
> it takes around 380ms to rename 221MB file
```
real	0m0.038s
user	0m0.012s
sys	0m0.006s
```
380 milliseconds is a lot. But isn't `0m0.038s` 38 milliseconds rather than 380?

Also, I ssh-ed into a pod, and timed python execution of a dummy script that does not rename a file, to account for a cold startup time:

```
time python -c 'import os'

real    0m0.017s
user    0m0.014s
sys     0m0.003s
```

Does it mean that the actual renaming of a file takes 38 - 17, or roughly 20 milliseconds?

Reply:
In my computer (average of 1000 repeats):
Renaming a 1-byte file: 0.68 milliseconds
Renaming a 500MB file: 0.66 ms (changing the file type)
Creating a 1-byte file: 1.3 ms

Reply:
@azangru  Yes you are right its 38ms. That was at last night 9pm I was a running that script probably in sleepy mode ;)

Reply:
> In my computer (average of 1000 repeats):

We should probably consider containers, nfs, and such. But I agree with you that it doesn't seem to be a time-consuming operation.

Reply:
Yes, the numbers will have a multiplier in k8s but it indicates that file renaming is one of the fastest file operations one can do (and the file size has no effect) so it should to be a non-issue for us. Which, I think, leaves this as our new best option for handling the temp file name (instead of using custom headers).

Reply:
@veidenberg I ran that script in k8s with NFS so it is as realistic as it can get.

Reply:
@veidenberg , I don't understand what is wrong with using custom headers? 
We already have file name after we receive the first chunk of the body and we have discussed this in code-review, What is the benefit of adding this completely unnecessary operation which can cause issues in file type, data , ... ?

Reply:
@Mehrnaz-Charkhchi I would ask the same questions but turn them the other way around:
- What's the point of adding a custom header when file name information already exists in the payload?
- Why add extra complexity to the payload?
- Why add a non-standard header to the request?